src/Processors/ISource.cpp

```diff
@@ -85,6 +87,18 @@ std::optional<ISource::ReadProgress> ISource::getReadProgress()
     return ReadProgress{res_progress, empty_limits};
 }
 
+void ISource::addTotalRowsApprox(size_t value)
+{
+    std::lock_guard lock(read_progress_mutex);
```

Comment:
@Algunenano Could you elaborate why the mutex is necessary? The `ISource` interface is not supposed to be accessed concurrently.

Reply:
You can see a sample report in https://github.com/ClickHouse/ClickHouse/pull/53050#issuecomment-1671355082 were ISource::work and ISource::progress are being called concurrently.

Reply:
Ok, but wasn't it introduced by this PR here? - https://github.com/ClickHouse/ClickHouse/pull/53050/files#diff-391fc60bfa23a1270bda630bb11853d9d50915a10b6bb66b2d116a3e1a287e02R209

Reply:
I'm not sure about what's wrong with the approach and I'm AFK for a few weeks. If you think it's the wrong approach we can revert and revisit later